source/extensions/http/cache/file_system_http_cache/cache_eviction_thread.cc

```diff
@@ -0,0 +1,75 @@
+#include "source/extensions/http/cache/file_system_http_cache/cache_eviction_thread.h"
+
+#include "source/common/api/os_sys_calls_impl.h"
+#include "source/extensions/http/cache/file_system_http_cache/file_system_http_cache.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Cache {
+namespace FileSystemHttpCache {
+
+CacheEvictionThread::CacheEvictionThread(Thread::ThreadFactory& thread_factory)
+    : os_sys_calls_(Api::OsSysCallsSingleton::get()),
+      thread_(thread_factory.createThread([this]() { work(); })) {}
+
+void CacheEvictionThread::addCache(FileSystemHttpCache& cache) {
+  absl::MutexLock lock(&cache_mu_);
+  bool inserted = caches_.emplace(&cache).second;
+  ASSERT(inserted);
+}
+
+void CacheEvictionThread::removeCache(FileSystemHttpCache& cache) {
+  absl::MutexLock lock(&cache_mu_);
+  bool removed = caches_.erase(&cache);
+  ASSERT(removed);
+  if (caches_.empty()) {
+    terminate();
+    thread_->join();
```

Comment:
I feel like it would be more natural to leave the cache thread running in case a cache is added after the last one is removed, and do the join() in the destructor for CacheEvictionThread, where I'd also assert caches_.empty(). WDYT?

The way you have it, if you own a CacheEvictionThread, the following sequence will fail:
```
  CacheEvictionThread cache_eviction_thread(thread_factory);
  FileSystemHttpCache cache;
  cache_eviction_thread.addCache(cache);
  cache_eviction_thread.removeCache(cache);
  cache_eviction_thread.addCache(cache);
  cache_eviction_thread.removeCache(cache);
```

Reply:
Or if you want to manage this with a singleton (which I generally prefer to avoid but that was already decided I guess) then I think I wouldn't start the thread in the constructor; I'd start the thread when the number of caches transitions from 0->1, and join it when the number of caches transitions from 1->0, and note this can happen multiple times during the lifetime of the system.

It might also be fine to just leave the thread running until shutdown.

Reply:
The reason I terminate the thread on removeCache rather than destructor is that the thread from which the destructor is called differs, which is apparently not an okay place to do thread-joining. Good point about symmetry though, so I've put thread create/terminate+join in addCache/removeCache.

Reply:
I can see why it might be a bit confusing to join from a different thread than the creator, but I don't see it being invalid in the C++ spec. Is there an Envoy standard?

It *is* UB to join from more than one thread though.

Moreover, I don't see how your change keeps us away from that situation.

Would it be easier to just never join the thread until the singleton is cleaned up? It would just be in a state when the thread was blocked on an Await that never comes. Of course we'd want to lazy-create in that case to avoid using a thread just cause this is linked in.